sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -439,6 +439,40 @@ int coap_get_blockopt(coap_pkt_t *pkt, uint16_t option, uint32_t *blknum, uint8_
     return (blkopt & 0x8) ? 1 : 0;
 }
 
+bool coap_find_uri_query(coap_pkt_t *pkt, const char *key, const char **value, size_t *len)
+{
+    uint8_t *opt_pos = NULL;
+
+    while (1) {
+        int len_key;
+        const void *key_data = coap_iterate_option(pkt, COAP_OPT_URI_QUERY,
+                                                   &opt_pos, &len_key);
+        if (key_data == NULL) {
+            return false;
+        }
+
+        const char *value_data = memchr(key_data, '=', len_key);
+
+        if (memcmp(key, key_data, value_data ? value_data - (char *)key_data : len_key)) {
+            continue;
+        }
+
+        if (value) {
+            *value = value_data
+                   ? value_data + 1
+                   : NULL;
+        }
+        if (len) {
+            *len = value_data
+                 ? len_key - (*value - (char *)key_data)
+                 : 0;
+        }
```

Comment:
I'm not a big fan of those rather obscure calculations, maybe you could simplify a bit with helper variables such as `len_value`?